File: chess/chess.cpp
/*
https://open.kattis.com/problems/chess
*/

#include <iostream>
#include <string>
#include <vector>
#include <math.h>   // sqrt
#include <stdlib.h> // abs

using namespace std;

struct Point
{
  public:
    Point(char x, char y)
        : x(x), y(y)
    {
    }
    Point()
    {
    }
    char x;
    char y;
};

float GetDistance(Point a, Point b)
{
    int x = abs(((int)a.x - 65) - ((int)b.x - 65));
    int y = abs(((int)a.y - 48) - ((int)b.y - 48));
    return sqrt(x * x + y * y);
}

vector<Point> GetPossibleMoves(Point pos)
{
    vector<Point> candidates, result;
    for (int i = 1; i < 8; i++)
    {
        candidates.push_back(Point((int)pos.x + 1 * i, (int)pos.y + 1 * i));
        candidates.push_back(Point((int)pos.x + 1 * i, (int)pos.y - 1 * i));
        candidates.push_back(Point((int)pos.x - 1 * i, (int)pos.y + 1 * i));
        candidates.push_back(Point((int)pos.x - 1 * i, (int)pos.y - 1 * i));
    }
    for (int i = 0; i < candidates.size(); i++)
    {
        if (candidates[i].x < 'A' || candidates[i].x > 'H')
            continue;
        if (candidates[i].y < '1' || candidates[i].y > '8')
            continue;
        result.push_back(candidates[i]);
    }
    return result;
}

vector<Point> Solve(Point pos, Point goal, vector<Point> path)
{
    if (path.size() > 8)
        return vector<Point>();
    float dist = GetDistance(pos, goal);
    if (dist == 0)
        return path;
    Point bestPos = Point(' ',' ');
    float bestDist = 999;
    vector<Point> candidates = GetPossibleMoves(pos);
    for (int i = 0; i < candidates.size(); i++)
    {
        Point newPos = candidates[i];
        float newDist = GetDistance(newPos, goal);
        if (newDist < bestDist)
        {
            bestPos = newPos;
            bestDist = newDist;
        }
    }
    vector<Point> newPath;
    for (int i = 0; i < path.size(); i++)
        newPath.push_back(path[i]);
    newPath.push_back(bestPos);
    return Solve(bestPos, goal, newPath);
}

int main(int argc, char *argv[])
{
    int n;
    cin >> n;

    for (int i = 0; i < n; i++)
    {
        vector<Point> result;
        vector<Point> path;

        Point a, b;

        cin >> a.x >> a.y;
        cin >> b.x >> b.y;

        path.push_back(a);
        result = Solve(a, b, path);

        if (result.size() == 0)
        {
            cout << "Impossible" << endl;
        }
        else
        {
            cout << result.size() - 1 << " ";
            for (int i = 0; i < result.size(); i++)
                cout << result[i].x << " " << result[i].y << " ";
            cout << endl;
        }
    }
}

File: chess/chess.py
"""
https://open.kattis.com/problems/chess
"""

import sys
from math import sqrt


def solve(pos, goal, walk):
    if len(walk) > 8:
        return None
    dist = get_distance(pos, goal)
    if dist == 0:
        return walk
    best_pos = None
    best_dist = 999
    for new_pos in get_possible_moves(pos):
        new_dist = get_distance(new_pos, goal)
        if new_dist < best_dist:
            best_pos = new_pos
            best_dist = new_dist
    return solve(best_pos, goal, walk + best_pos[0] + best_pos[1])


def get_possible_moves(pos):
    valid_x = ("A", "B", "C", "D", "E", "F", "G", "H")
    valid_y = ("1", "2", "3", "4", "5", "6", "7", "8")
    for d in [(1, 1), (1, -1), (-1, 1), (-1, -1)]:
        for i in range(1, 8):
            new_pos = (chr(ord(pos[0]) + d[0] * i),
                       str(int(pos[1]) + d[1] * i))
            if not new_pos[0] in valid_x:
                continue
            if not new_pos[1] in valid_y:
                continue
            yield new_pos


def get_distance(a, b):
    x = abs((ord(a[0]) - 65) - (ord(b[0]) - 65))
    y = abs(int(a[1]) - int(b[1]))
    return sqrt(x**2 + y**2)


def main():
    n = int(sys.stdin.readline())
    for i in range(n):
        arr = sys.stdin.readline().split()
        a = (arr[0], arr[1])
        b = (arr[2], arr[3])
        moves = solve(a, b, a[0] + a[1])
        if moves is None:
            print("Impossible")
        else:
            print(len(moves) // 2 - 1, end=' ')
            for c in moves:
                print(c, end=' ')
            print()


if __name__ == "__main__":
    main()
